python/tempo/interpol.py

```diff
@@ -0,0 +1,321 @@
+        :param ts_col   - timestamp column name
+        :param partition_cols  - partition columns names
+        """
+        # Validate parameters
+        self.__validate_fill(fill)
+        self.__validate_col(tsdf.df, partition_cols, target_cols, ts_col)
+
+        # Resample and Normalize Columns
+        # TODO: Resampling twice, possible optimization is to see if we can do it only resample once
+        no_fill: DataFrame = tsdf.resample(
+            freq=sample_freq, func=sample_func, metricCols=target_cols
+        ).df
+
+        # Build columns list to joining the two sampled series
+        join_cols: List[str] = partition_cols + [ts_col]
+
+        # Get series zero filled
+        zero_fill: DataFrame = tsdf.resample(
+            freq=sample_freq, func=sample_func, fill=True, metricCols=target_cols
+        ).df.select(*join_cols)
+
+        # Join sampled DataFrames - generate complete timeseries
+        joined_series: DataFrame = zero_fill.join(
+            no_fill,
+            join_cols,
+            "left",
+        )
+
+        # Perform interpolation on each target column
+        output_list: List[DataFrame] = []
+        for target_col in target_cols:
+            # Mark columns that are interpolated
+            marked_series: DataFrame = joined_series.withColumn(
+                f"is_{target_col}_interpolated",
+                when(col(target_col).isNull(), True).otherwise(False),
+            )
+
+            # Add surrogate ts_col to get missing values
+            with_surrogate_ts: DataFrame = marked_series.withColumn(
+                "surrogate_ts",
+                when(col(target_col).isNull(), None).otherwise(col(ts_col)),
+            )
+
+            # # Normalize Timestamp to Epoch Seconds
+            epoch_series: DataFrame = with_surrogate_ts.withColumn(
+                ts_col, unix_timestamp(ts_col)
+            ).withColumn("surrogate_ts", unix_timestamp("surrogate_ts"))
+
+            # # Interpolate target columns
+            interpolated_result: DataFrame = self.__interpolate_column(
+                epoch_series, partition_cols, ts_col, target_col, fill
+            )
+
+            # Denormalize output
+            columns_to_drop: List[str] = [
+                "readvalue_ff",
+                "readvalue_bf",
+                "readtime_ff",
+                "readtime_bf",
+                "surrogate_ts",
+            ] + target_cols
+            columns_to_drop.remove(target_col)
+            normalized_output: DataFrame = (
+                # Drop intermediatry columns
+                interpolated_result.drop(*columns_to_drop)
+                # Convert timestamp column back to timestamp
+                .withColumn(ts_col, to_timestamp(ts_col))
+            )
+            output_list.append(normalized_output)
+
+        # Join output_list into single DataFrame to output
+        output: DataFrame = output_list[0]
+        output.cache()
```

Comment:
A memory cache will be expensive here. Instead of using this, I would rely on Delta Lake - we can make a comment that this step is optimized on NVMe-supported cloud VMs.

Reply:
Removed `output.cache()`